src/Compiler/Checking/PatternMatchCompilation.fs

```diff
@@ -1669,7 +1669,7 @@ let CompilePatternBasic
         let used = HashSet<_>(accTargetsOfDecisionTree dtree [], HashIdentity.Structural)
 
         clauses |> List.iteri (fun i c ->
-            if not (used.Contains i) then warning (RuleNeverMatched c.Range))
+            if not (used.Contains i) then warning (RuleNeverMatched c.Pattern.Range))
```

Comment:
Would it maybe work to use `c.Range |> withEnd c.Pattern.Range` (using range helpers added recently in https://github.com/dotnet/fsharp/pull/15700)?
I mean, to get the range that Petr wants, which I also think would be a bit clearer

Reply:
Yeah looks like this is some kind of a rabbit hole. Thanks for your grit :)

Reply:
@abonie @psfinaki Updated the range calculation and now covers all the cases I could think of(See tests).

Reply:
@edgarfgp this is absolutely fascinating job!

To match your efforts (pun intended), I manually verified all the tests cases you added :D This is definitely sufficient IMO, we might discover more interesting cases later but we have a solid testing framework for this anyway.

Thank you!

Reply:
Thanks. I see plenty of opportunities to improve error reporting. So this will be my focus and legacy for F# 🙂